### PR TITLE
(🎁) Support Python 3.11

### PR DESCRIPTION
**Pull request by KotlinIsland**
- [ ] Closes #xxxx (Replace xxxx with the Github issue number)
- [ ] Tests added: Please use `assert_type()` to assert the type of any return value

### CODE REVIEW

**Comment by bashtage** [comment]
Need to wait for lxml to have 3.11 wheels.

**Comment by KotlinIsland** [comment]
https://github.com/lxml/lxml/pull/355
https://github.com/lxml/lxml/pull/356

**Comment by bashtage** [comment]
There are 3.11 wheels on PyPu if 1.5.1, so it looks like 3.11 is supported by pandas.

**Comment by BrenoJesusFernandes** [comment]
Be careful, looks like we still don't have  a stable version for python 3.11 for macos in github actions: https://github.com/actions/python-versions/releases/tag/3.11.0-3320066239

**Comment by Dr-Irv** [comment]
CI is failing because no stable version of python 3.11 for MacOS yet and no `pyarrow` for python 3.11

**Comment by Dr-Irv** [comment]
Issue is still that `pyarrow` doesn't work with python 3.11.  The CI sees no binary, then tries to build the wheel, and it fails.

There might be a workaround here to skip the requirement of pyarrow when testing on 3.11 (and then to adjust the tests accordingly to do a `pytest.skip`) but I don't know how to modify `pyproject.toml` to do that.

@BrenoJesusFernandes Any ideas on how to make the poetry requirement for `pyarrow` conditional on python <= 3.10 in `pyproject.toml` ?

**Comment by Dr-Irv** [comment]
spoke with someone who is familiar with `pyarrow`, and they think 3.11 wheels will be available within a week, so maybe easiest to wait for that

**Comment by BrenoJesusFernandes** [comment]
I don't get why pandas is available in python 3.11, when one requirement is not available. It's seems wrong for me. In poetry last versio, you can define multiples dependencies besides the dev-dependecies, but I think we should wait the dependecies to be available in 3.11.

**Comment by Dr-Irv** [comment]
> I don't get why pandas is available in python 3.11, when one requirement is not available. It's seems wrong for me. In poetry last versio, you can define multiples dependencies besides the dev-dependecies, but I think we should wait the dependecies to be available in 3.11.

`pyarrow` is an _optional_ dependency.  So that's why `pandas` is available for python 3.11, because you can use `pandas` with python 3.11, you just can't use `pyarrow` with it.

So if there is a way to define the dependencies as one set for python <= 3.10 that we have now, and another set that is the same, except removes `pyarrow` for python 3.11, we can move forward.  Then we modify the tests that use `pyarrow` to test if `pyarrow` is available.  That's how the tests for pandas with python 3.11 are set up.

Or we wait a week (hopefully) until  `pyarrow` builds wheels for python 3.11

**Comment by Dr-Irv** [comment]
> is there some test using pyarrow? if exits tests like that it is gonna fail in CI.

There are a couple of tests in `test_pandas.py` and `test_frame.py`.  They are protected with a `pytest.importorskip` .  

So we just need to modify the `pyproject.toml` file to indicate that `pyarrow` should not be installed when testing python 3.11.  Then the tests should work.

**Comment by BrenoJesusFernandes** [comment]
hmm I see, so it's simple than I thought, just add something like that:

[tool.poetry.dev-dependencies]
pyarrow = { version = "currently-version", python = "<3.11" }
]

**Comment by Dr-Irv** [comment]
@KotlinIsland if you want to try what @BrenoJesusFernandes suggests above, that might make the CI work.

**Comment by bellini666** [comment]
Since this is a stubs package, does it make sense to limit the upper bound of the python version? I would say that even libs shouldn't do that themselves, unless there's a very good reason.

This is very limiting for people using poetry for example due to [this issue](https://github.com/python-poetry/poetry/issues/697)

**Comment by Dr-Irv** [comment]
> Since this is a stubs package, does it make sense to limit the upper bound of the python version? I would say that even libs shouldn't do that themselves, unless there's a very good reason.

The limit is there for now because we can't run a successful test with python 3.11.  We're not willing to release the stubs for a python version unless we can successfully test the stubs on that version.

In the past, we have had issues with the stubs being compatible with one version of python, but not another.  So our tests help make sure we are not making that mistake.

**Comment by KotlinIsland** [comment]
Looks like `tables` and `lxml` don't have 3.11 wheels either.

**Comment by KotlinIsland** [comment]
@Dr-Irv Okay, I have made `arrow`, `tables` and `lxml` conditional dependencies, and I have added a skip check for 3.11

File: tests/test_io.py
Comment (KotlinIsland) [reply tests/test_io.py:75]: Strange, it made the tests pass

Comment (Dr-Irv) [reply tests/test_io.py:75]: > This is the output of pytest, it seems to me as if `test_io` is in some way using `pyarrow`?

I'm guessing there are dependencies.  So `test_orc` pattern needs `pyarrow`, `test_xml` pattern needs `lxml`, `test_hdf` pattern needs `pytables`, etc.  So we should do the skips based on the particular test (and underlying library) being done.

Comment (KotlinIsland) [reply tests/test_io.py]: Hows that look?

**Comment by KotlinIsland** [comment]
Looks like there is some issues with mypy: https://github.com/python/mypy/issues/13627

This is causing the ci to fail for 3.11 and will require updating mypy to 0.982

**Comment by Dr-Irv** [comment]
@KotlinIsland Please see https://github.com/pandas-dev/pandas-stubs/pull/398#pullrequestreview-1161937232

We have to wait for mypy 0.990

**Comment by Dr-Irv** [comment]
> @KotlinIsland Please see [#398 (review)](https://github.com/pandas-dev/pandas-stubs/pull/398#pullrequestreview-1161937232)
> 
> We have to wait for mypy 0.990

@KotlinIsland We are now testing with `mypy` 0.990 via #419 so can you merge with that, resolve conflicts, and see if the python 3.11 support will work?

**Comment by KotlinIsland** [comment]
@Dr-Irv Can you trigger the CI?

**Comment by Dr-Irv** [comment]
Currently two issues (but there might be more):
In `test_pandas.py`, have to use a pattern like this:
```python
import pandas.util._test_decorators as td  # at top of file somewhere

pyarrow_skip = td.skip_if_no("pyarrow")


@pyarrow_skip
def test_arrow_dtype() -> None:
    import pyarrow as pa

    check( ...
```
and in `pandas-stubs/core/arrays/arrow/dtype.pyi` (somewhere near the top):
```python
if sys.version_info < (3, 11):
    import pyarrow as pa
else:
    pa = Any
```

I'd also suggest setting up a local environment with python 3.11 and running the tests locally instead of waiting for me or someone else to do a CI approval.

**Comment by Dr-Irv** [comment]
I played around a bit with this, and uncovered a pyright bug that doesn't ignore imports based on the python version test.  https://github.com/microsoft/pyright/issues/4182

So we have to wait for that to get fixed to be able to do conditional imports.

File: tests/test_pandas.py
Comment (Dr-Irv) [inline tests/test_pandas.py:30]: Shouldn't this be 
```python
pyarrow_skip = pytest.mark.skipif(sys.version_info >= (3, 11), "pyarrow is not available for 3.11 yet")
"""This is only needed temporarily due to no wheels being available for arrow on 3.11"""
```

File: tests/test_pandas.py
Comment (Dr-Irv) [inline tests/test_pandas.py:27]: this import is failing.

**Review by Dr-Irv — CHANGES_REQUESTED** [review]

File: tests/test_io.py
Comment (bashtage) [inline tests/test_io.py:76]: Wrong reason in the explanation here. Should be lxml not arrow.

File: tests/test_pandas.py
Comment (bashtage) [inline tests/test_pandas.py:465]: These is there for a reason to protect Windows.  I think it should stay.

File: tests/test_pandas.py
Comment (bashtage) [inline tests/test_pandas.py:30]: Appears to be not needed.

**Review by bashtage — CHANGES_REQUESTED** [review]
Some fixes needed.

**Comment by Dr-Irv** [comment]
I created #435 to move this along.  Took some of your ideas from here.  Had to handle an issue with pandas that's a bit thorny and want to see if I can coordinate release of this with release of pandas 1.5.2 if a PR there is included.

Thanks for working on this.  It was a great head start.

**Pull request closed without merging**
